# Minimal RLOO Fine-Tuning for Code Tasks (MBPP → HumanEval)

This notebook shows a **minimal end‑to‑end** workflow to fine‑tune a small code model with **RLOO** (Reinforcement Learning with Leave‑One‑Out) on **MBPP (sanitized)** and then **evaluate on HumanEval**.

**Default model:** `HuggingFaceTB/SmolLM2-360M-Instruct` (tiny & fast).  
**Swap-in option:** Any compatible HF model id (e.g., `Qwen/Qwen2.5-Coder-1.5B-Instruct`).

> ⚠️ **Security note**: This workflow **executes model‑generated Python** when scoring with unit tests. Run in a controlled environment (e.g., Docker/VM). You can also use `evalplus` which isolates execution. 


In [2]:
# If running locally, uncomment the next line to ensure fresh packages.
%pip install -U transformers accelerate datasets trl peft safetensors human-eval evalplus bitsandbytes

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 1.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 1.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ...   Downloading fire-0.7.1-py3-none-any.whl.metadata (5.8 kB)
  Preparing metadata (setup.py) ... done
done
  Preparing metadata (setup.py) ... Collecting tempdir>=0.7.1 (from evalplus)
  Preparing metadata (setup.py) ... done
done
  Preparing metadata (setup.py) ... Collecting stop-sequencer>=1.2.3 (from evalplus)
  Preparing metadata (setup.py) ... done
done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0.0/12.0 MB ? eta -:--:--Downloading transformers-4.57.1-py3-none-any.whl (12.0 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 96.0 MB/s eta 0:00:00:00:01:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 96.0 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 375.8/375.8 kB 19.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 462.8/462.8 kB 17

In [2]:
# Clean up TensorFlow packages to avoid import errors (not needed for this torch-only workflow)
%pip uninstall -y tensorflow tensorflow-intel tensorflow-io-gcs-filesystem

Note: you may need to restart the kernel to use updated packages.


In [3]:
import importlib.util
import os
print("tensorflow spec:", importlib.util.find_spec("tensorflow"))
print("HF_TOKEN present:", bool(os.environ.get("HF_TOKEN")))

tensorflow spec: ModuleSpec(name='tensorflow', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7871670ec550>, origin='/usr/local/lib/python3.11/dist-packages/tensorflow/__init__.py', submodule_search_locations=['/usr/local/lib/python3.11/dist-packages/tensorflow'])
HF_TOKEN present: False


In [4]:
from huggingface_hub import HfFolder
hf_keys = [k for k in os.environ if k.startswith("HF")]
print("Env keys starting with 'HF':", hf_keys)
print("HF_TOKEN present:", bool(os.environ.get("HF_TOKEN")))
print("HfFolder token present:", bool(HfFolder.get_token()))

Env keys starting with 'HF': []
HF_TOKEN present: False
HfFolder token present: False


In [5]:
import evalplus.data.mbpp as mbpp_mod
import inspect
print([name for name, obj in inspect.getmembers(mbpp_mod) if inspect.isfunction(obj)])

['_ready_mbpp_plus_path', 'completeness_check', 'get_dataset_metadata', 'get_mbpp', 'get_mbpp_plus', 'get_mbpp_plus_hash', 'make_cache', 'mbpp_deserialize_inputs', 'mbpp_serialize_inputs', 'stream_jsonl']


In [6]:
from evalplus.data.mbpp import get_mbpp
import itertools
mbpp_evalplus = get_mbpp()
print("evalplus size:", len(mbpp_evalplus))
first_key = next(iter(mbpp_evalplus))
print("Sample key:", first_key)
print(mbpp_evalplus[first_key].keys())

evalplus size: 427
Sample key: 2
dict_keys(['source_file', 'task_id', 'prompt', 'code', 'test_imports', 'test_list'])
evalplus size: 427
Sample key: 2
dict_keys(['source_file', 'task_id', 'prompt', 'code', 'test_imports', 'test_list'])


In [3]:
import os, re, json, math, textwrap, multiprocessing as mp, queue, signal, sys
from dataclasses import dataclass
from pathlib import Path

# Disable optional back ends we do not need (avoids pulling in TensorFlow/Flax)
os.environ["TRANSFORMERS_NO_TF"] = "1"
os.environ["TRANSFORMERS_NO_FLAX"] = "1"
os.environ["USE_TF"] = "0"
os.environ["USE_FLAX"] = "0"
os.environ.setdefault("HF_HUB_DISABLE_TELEMETRY", "1")

from transformers.utils import import_utils as _hf_import_utils
_hf_import_utils.is_tf_available = lambda: False
# also update the public shortcut to be safe
from transformers import utils as _hf_utils
_hf_utils.is_tf_available = lambda: False

import torch
from datasets import load_dataset, Dataset
from transformers import AutoTokenizer, AutoModelForCausalLM
from trl import RLOOConfig, RLOOTrainer
from peft import LoraConfig

# ---- Choose your base model
MODEL_NAME = os.environ.get("BASE_MODEL", "HuggingFaceTB/SmolLM2-360M-Instruct")
# Examples:
#   "HuggingFaceTB/SmolLM2-360M-Instruct"  (≈360M)
#   "Qwen/Qwen2.5-Coder-1.5B-Instruct"     (≈1.5B)
#   "Qwen/Qwen2.5-Coder-7B-Instruct"       (heavy; consider 8-bit)

# Generation defaults for RLOO and eval
GEN_K = int(os.environ.get("RLOO_NUM_GENERATIONS", 4))     # leave-one-out over K completions
MAX_PROMPT_LEN = 512
MAX_COMPLETION_LEN = 256

DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using device: {DEVICE}")


/usr/local/lib/python3.11/dist-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` 

Using device: cuda


## Load MBPP (sanitized) and build prompts
We’ll use the **sanitized** split (427 items). It includes a problem prompt, canonical solution, and unit tests. For reliable function names, we **extract only the function signature** from the canonical solution and add it to the prompt (no leakage of solution body).

In [7]:
from huggingface_hub import list_repo_files
from huggingface_hub.utils import RepositoryNotFoundError
hf_token = os.environ.get("HF_TOKEN")
try:
    # Explicitly mark this repo as a dataset; otherwise the hub looks under models and fails.
    files = list_repo_files("Muennighoff/mbpp", repo_type="dataset", token=hf_token)
    print([f for f in files if "sanitized" in f][:20])
except RepositoryNotFoundError:
    print("Cannot list 'Muennighoff/mbpp'. Accept the dataset terms and set HF_TOKEN first.")


['data/sanitized-mbpp.json']


In [4]:
import os
# Configure this session to use the larger Qwen coder checkpoint.
os.environ["BASE_MODEL"] = "Qwen/Qwen2.5-Coder-1.5B-Instruct"
# Make sure GPUs remain visible; Kaggle exposes two T4s when enabled.
os.environ["CUDA_VISIBLE_DEVICES"] = os.environ.get("CUDA_VISIBLE_DEVICES", "0,1")
# Disable Weights & Biases auto-login so the demo doesn't block on auth prompts.
os.environ["WANDB_DISABLED"] = "true"
# Keep the compile/demo run light for local validation.
os.environ["MBPP_TRAIN_ITEMS"] = "32"
os.environ["RLOO_MAX_STEPS"] = "3"
os.environ["RLOO_NUM_GENERATIONS"] = "2"
print("BASE_MODEL set to:", os.environ["BASE_MODEL"])
print("CUDA_VISIBLE_DEVICES:", os.environ.get("CUDA_VISIBLE_DEVICES"))
print("MBPP_TRAIN_ITEMS:", os.environ.get("MBPP_TRAIN_ITEMS"))
print("RLOO_MAX_STEPS:", os.environ.get("RLOO_MAX_STEPS"))
print("RLOO_NUM_GENERATIONS:", os.environ.get("RLOO_NUM_GENERATIONS"))

BASE_MODEL set to: Qwen/Qwen2.5-Coder-1.5B-Instruct
CUDA_VISIBLE_DEVICES: 0,1
MBPP_TRAIN_ITEMS: 32
RLOO_MAX_STEPS: 3
RLOO_NUM_GENERATIONS: 2


In [6]:
from evalplus.data.mbpp import get_mbpp

# Load sanitized MBPP via evalplus (avoids gated Hugging Face script)
mbpp_dict = get_mbpp()
mbpp = Dataset.from_list(list(mbpp_dict.values()))
mbpp = mbpp.shuffle(seed=42)
print(mbpp[0].keys())
# expected fields: 'prompt', 'code', 'test_imports', 'test_list', 'challenge_test_list'

# --- helpers to build a good training prompt ---
DEF_RE = re.compile(r"^\s*def\s+\w+\s*\(.*\)\s*:")

def extract_signature(code: str) -> str:
    for line in code.splitlines():
        if DEF_RE.match(line):
            return line.strip()
    return "def solution():\n    pass"


def build_training_prompt(rec):
    sig = extract_signature(rec["code"])  # only the signature, no body
    base = rec["prompt"].strip()
    prompt = f"""
You are a Python coding assistant.
Write a **correct and efficient** solution that **defines exactly** this function signature and passes the hidden tests.

# Problem
{base}

# Function signature (must match exactly)
{sig}

# Output format
Return **only valid Python code** implementing the function. **No** markdown, comments, prints, or extra text.
""".strip()
    return prompt

# Build a compact training set for a quick run
train_size = int(os.environ.get("MBPP_TRAIN_ITEMS", 150))  # try 150 for a few minutes; increase for better results
subset = mbpp.select(range(train_size))

train_records = subset.map(lambda r: {
    "prompt_text": build_training_prompt(r),
    "test_setup": "\n".join(r.get("test_imports", []) or []),
    "tests": r.get("test_list", []) or [],
})

print("Example data:\n")
print("Prompt text:\n")
print(train_records[0]["prompt_text"])
print("\nTest setup:\n")
print(train_records[0]["test_setup"])
print("\nTests:\n")
print(train_records[0]["tests"])


dict_keys(['source_file', 'task_id', 'prompt', 'code', 'test_imports', 'test_list'])
dict_keys(['source_file', 'task_id', 'prompt', 'code', 'test_imports', 'test_list'])


Map:   0%|          | 0/32 [00:00<?, ? examples/s]

Example data:

Prompt text:

You are a Python coding assistant.
Write a **correct and efficient** solution that **defines exactly** this function signature and passes the hidden tests.

# Problem
Write a function to find the shared elements from the given two lists.

# Function signature (must match exactly)
def similar_elements(test_tup1, test_tup2):

# Output format
Return **only valid Python code** implementing the function. **No** markdown, comments, prints, or extra text.

Test setup:



Tests:

['assert set(similar_elements((3, 4, 5, 6),(5, 7, 4, 10))) == set((4, 5))', 'assert set(similar_elements((1, 2, 3, 4),(5, 4, 3, 7))) == set((3, 4))', 'assert set(similar_elements((11, 12, 14, 13),(17, 15, 14, 13))) == set((13, 14))']


In [5]:
import torch
print("torch.cuda.is_available():", torch.cuda.is_available())
print("GPU count:", torch.cuda.device_count())
if torch.cuda.is_available():
    for idx in range(torch.cuda.device_count()):
        print(f"GPU {idx}:", torch.cuda.get_device_name(idx))

torch.cuda.is_available(): True
GPU count: 2
GPU 0: Tesla T4
GPU 1: Tesla T4


## Reward function
The reward is **fraction of unit tests passed** (0.0–1.0) for each generated completion.  
We execute code in a **separate process with a timeout** for basic isolation. For stricter isolation, consider running this notebook inside Docker and/or using `evalplus` for evaluation time.

In [7]:
def _mp_worker(q: mp.Queue, code: str, tests: list[str], setup: str):
    """Run student code plus tests in isolation and push (passed, total) to q."""
    try:
        glb = {}
        if setup:
            exec(setup, glb, glb)
        exec(code, glb, glb)
        passed = 0
        total = len(tests)
        for t in tests:
            try:
                exec(t, glb, glb)
                passed += 1
            except Exception:
                pass
        q.put((passed, total))
    except Exception:
        q.put((0, len(tests)))


def _exec_in_subproc(code: str, tests: list[str], setup: str, timeout_s: float = 3.0):
    """Execute in a short-lived process with a timeout; return (passed, total)."""
    q = mp.Queue()
    p = mp.Process(target=_mp_worker, args=(q, code, tests, setup))
    p.start()
    p.join(timeout_s)
    if p.is_alive():
        p.terminate()
        return 0, len(tests)
    try:
        return q.get_nowait()
    except queue.Empty:
        return 0, len(tests)

# Match fenced code blocks so we can pull out raw Python when the model returns markdown.
CODE_FENCE_RE = re.compile(r"```(?:python)?\n(.*?)```", re.DOTALL)

def extract_code_only(txt: str) -> str:
    """Strip markdown fences or stray backticks; fall back to raw text."""
    m = CODE_FENCE_RE.search(txt)
    if m:
        return m.group(1).strip()
    return txt.strip().strip("`")

# TRL reward function signature: it receives completions alongside dataset columns.

def mbpp_reward(completions, tests, test_setup, **kwargs):
    """Score each completion by fraction of MBPP tests that pass."""
    outs = []
    for comp, ts, setup in zip(completions, tests, test_setup):
        # TRL may give chat-format responses (list of role/content dicts); normalize to plain text.
        if isinstance(comp, list) and len(comp) and isinstance(comp[0], dict) and "content" in comp[0]:
            comp_text = comp[0]["content"]
        else:
            comp_text = str(comp)
        code = extract_code_only(comp_text)
        passed, total = _exec_in_subproc(code, ts, setup, timeout_s=3.0)
        outs.append(0.0 if total == 0 else float(passed) / float(total))
    return outs

# quick self-check on one item
_test_code = """
"""
sc = train_records[0]
_score = mbpp_reward([
    """def noop():\n    return 0"""
], tests=[sc["tests"]], test_setup=[sc["test_setup"]])
print("example reward:", _score)


example reward: [0.0]


## Load tokenizer/model & configure RLOO (LoRA)
We keep it small and cheap: LoRA adapters on a tiny model. Increase steps / dataset size later.

In [8]:
# Tokenizer drives both prompt encoding and generation decoding during training/eval.
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME, use_fast=True)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token

# LoRA adapter hyperparameters control how much of the attention/MLP blocks we fine-tune.
peft_cfg = LoraConfig(
    r=8,  # rank of the low-rank adapters; higher values = more trainable params
    lora_alpha=16,  # scales the adapter update; affects learning capacity
    lora_dropout=0.05,  # regularizes adapter activations to curb overfitting
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=["q_proj","k_proj","v_proj","o_proj","gate_proj","up_proj","down_proj"],
)

# Core RLOO loop settings determine optimization stability, sampling cost, and logging cadence.
rloo_cfg = RLOOConfig(
    output_dir="./smollm2-mbpp-rloo",  # checkpoint + log destination reused by later save/load
    learning_rate=2e-5,  # step size for adapter weights; too high destabilizes RLOO updates
    per_device_train_batch_size=2,  # number of prompts per device; limited by VRAM during sampling
    gradient_accumulation_steps=1,  # scales effective batch; increase to smooth rewards if memory allows
    num_generations=GEN_K,  # completions per prompt; defines leave-one-out pool and runtime cost
    generation_batch_size=GEN_K,  # how many generations sampled concurrently; impacts GPU RAM footprint
    max_prompt_length=MAX_PROMPT_LEN,  # truncation guard so prompts fit into context window
    max_completion_length=MAX_COMPLETION_LEN,  # cap on generated tokens; affects sampling time/test cost
    beta=0.01,  # leave-one-out baseline strength; tunes variance reduction in RLOO objective
    logging_steps=5,  # tensorboard/console logging frequency
    save_steps=50,  # checkpoint interval that ties to checkpoints reused for eval
    max_steps=int(os.environ.get("RLOO_MAX_STEPS", 60)),  # total optimizer steps; governs training duration
    bf16=False,  # disable bf16 to stay compatible with CPU-only or older GPU runtimes
    fp16=False,  # explicitly stay in full precision on this hardware
)

# Trainer wires model loading, reward loop, dataset, and adapter config together.
trainer = RLOOTrainer(
    model=MODEL_NAME,  # base HF model id; governs architecture capacity and tokenizer compatibility
    args=rloo_cfg,
    reward_funcs=mbpp_reward,  # custom reward defined above; shapes gradient signal
    processing_class=tokenizer,  # ensures prompts/generations reuse the tokenizer configured here
    train_dataset=train_records,  # preprocessed MBPP subset
    peft_config=peft_cfg,  # attaches LoRA adapters with the settings defined above
)
trainer

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/655 [00:00<?, ?B/s]

Using the `WANDB_DISABLED` environment variable is deprecated and will be removed in v5. Use the --report_to flag to control the integrations used for logging result (for instance --report_to none).
/usr/local/lib/python3.11/dist-packages/trl/trainer/rloo_trainer.py:241: UserWarning: This trainer will soon be moved to trl.experimental and is a candidate for removal. If you rely on it and want it to remain, please share your comments here: https://github.com/huggingface/trl/issues/4223. Silence this warning by setting environment variable TRL_EXPERIMENTAL_SILENCE=1.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/trl/trainer/rloo_trainer.py:241: UserWarning: This trainer will soon be moved to trl.experimental and is a candidate for removal. If you rely on it and want it to remain, please share your comments here: https://github.com/huggingface/trl/issues/4223. Silence this warning by setting environment variable TRL_EXPERIMENTAL_SILENCE=1.
  warnings.warn(


config.json:   0%|          | 0.00/846 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/724M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/132 [00:00<?, ?B/s]

2025-11-07 13:14:42.834968: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1762521283.236724      77 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1762521283.353081      77 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


### Sanity check: generate before training

In [9]:
prompt0 = train_records[0]["prompt_text"]
inputs = tokenizer(prompt0, return_tensors="pt").to(trainer.accelerator.device)
input_len = inputs["input_ids"].shape[1]
with torch.no_grad():
    output_ids = trainer.model.generate(
        **inputs,
        max_new_tokens=192,
        do_sample=True,
        temperature=0.7,
        pad_token_id=tokenizer.pad_token_id,
    )
completion_ids = output_ids[0][input_len:]
completion = tokenizer.decode(completion_ids, skip_special_tokens=True).strip()
print("--- Prompt (truncated) ---")
print("\n".join(prompt0.splitlines()[:8]))
print("\n--- Sampled completion ---")
print(completion if completion else "[empty completion]")


--- Prompt (truncated) ---
You are a Python coding assistant.
Write a **correct and efficient** solution that **defines exactly** this function signature and passes the hidden tests.

# Problem
Write a function to find the shared elements from the given two lists.

# Function signature (must match exactly)
def similar_elements(test_tup1, test_tup2):

--- Sampled completion ---
# Test cases
test_test_tup1 = ([(1, "A"), (2, "B"), (3, "C")], [(4, "D"), (5, "E")])
test_test_tup2 = ([(6, "F"), (7, "G"), (8, "H")], [(9, "I"), (10, "J")])

# Output
assert similar_elements(test_test_tup1, test_test_tup2) == [("F", "G"), ("I", "J"), ("H", "I")]

assert similar_elements([("A", "A"), ("B", "B"), ("C", "C")], [("A", "A"), ("B", "B"), ("C", "C")]) == []

assert similar_elements([("


## Train (tiny demo run)
Crank up `RLOO_MAX_STEPS` and `MBPP_TRAIN_ITEMS` for real gains.

In [10]:
trainer.train()

/usr/local/lib/python3.11/dist-packages/torch/utils/checkpoint.py:87: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The 

TrainOutput(global_step=3, training_loss=0.0, metrics={'train_runtime': 72.6484, 'train_samples_per_second': 0.165, 'train_steps_per_second': 0.041, 'total_flos': 0.0, 'train_loss': 0.0})

In [ ]:
save_dir = rloo_cfg.output_dir
trainer.save_model(save_dir)
print("Saved to:", save_dir)

Saved to: ./smollm2-mbpp-rloo


In [13]:
sample_indices = list(range(3))
sampled = []
for idx in sample_indices:
    rec = train_records[idx]
    prompt = rec["prompt_text"]
    inputs = tokenizer(prompt, return_tensors="pt").to(trainer.accelerator.device)
    input_len = inputs["input_ids"].shape[1]
    with torch.no_grad():
        gen_ids = trainer.model.generate(
            **inputs,
            max_new_tokens=192,
            do_sample=True,
            temperature=0.7,
            top_p=0.95,
            pad_token_id=tokenizer.pad_token_id,
            eos_token_id=tokenizer.eos_token_id,
        )
    decoded = tokenizer.decode(gen_ids[0][input_len:], skip_special_tokens=True)
    decoded_raw = tokenizer.decode(gen_ids[0][input_len:], skip_special_tokens=False)
    completion = decoded.strip() or decoded_raw.strip()
    sampled.append({
        "idx": idx,
        "prompt": prompt,
        "completion": completion,
        "tests": rec["tests"],
        "setup": rec["test_setup"],
    })

rewards = mbpp_reward(
    [item["completion"] for item in sampled],
    tests=[item["tests"] for item in sampled],
    test_setup=[item["setup"] for item in sampled],
)

for item, reward in zip(sampled, rewards):
    preview = item["completion"][:200] + ("..." if len(item["completion"]) > 200 else "")
    print(f"Task {item['idx']}: reward={reward:.2f}")
    print("Prompt snippet:")
    print("\n".join(item["prompt"].splitlines()[:4]))
    print("Completion preview:")
    print(preview if preview else "[empty]")
    print("-" * 60)

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Task 0: reward=0.00
Prompt snippet:
You are a Python coding assistant.
Write a **correct and efficient** solution that **defines exactly** this function signature and passes the hidden tests.

# Problem
Completion preview:
As for the first step.

To the time.

The way to "```

```


As the set, and a 01.



When you are a few possible or I consider the number in your career, and you get the town and a day. If a few othe...
------------------------------------------------------------
Task 1: reward=0.00
Prompt snippet:
You are a Python coding assistant.
Write a **correct and efficient** solution that **defines exactly** this function signature and passes the hidden tests.

# Problem
Completion preview:
It are a way to get to the group of a group of 1 2.

It's work for it's a 2)

```




We are a list of the first step (4.





This can be a bit, and "In 122221-B = 000200, the other types of a key, 1...
------------------------------------------------------------
Task 2: reward=0.00
Prom

## Evaluate on HumanEval (pass@1)
We’ll create a `samples.jsonl` with one completion per task, then run the official HumanEval harness.

> On Windows, prefer `pip install human-eval-windows`.
> If you want more rigorous evaluation, see `evalplus` too.


In [ ]:
from datasets import load_dataset
from tqdm import tqdm

humaneval = load_dataset("openai/openai_humaneval", split="test")

# Reuse trainer.model/tokenizer in plain HF generate for speed
model = trainer.model
model.eval()

def make_completion(prompt: str, max_new_tokens=192, temperature=0.2):
    # Use plain text prompting (no chat template) to keep it simple
    inputs = tokenizer([prompt], return_tensors="pt").to(model.device)
    with torch.no_grad():
        gen_ids = model.generate(**inputs, max_new_tokens=max_new_tokens, do_sample=True,
                                 temperature=temperature, eos_token_id=tokenizer.eos_token_id)
    out = tokenizer.decode(gen_ids[0][inputs["input_ids"].shape[1]:], skip_special_tokens=True)
    return out.strip()

# Build HumanEval prompts: use the provided "prompt" field directly
samples_path = Path("./humaneval_samples.jsonl")
with samples_path.open("w", encoding="utf-8") as f:
    for row in tqdm(humaneval):
        prompt = row["prompt"]
        comp = make_completion(prompt)
        # strip code fences if present
        comp = re.sub(r"^```(?:python)?|```$", "", comp).strip()
        rec = {"task_id": row["task_id"], "completion": comp}
        f.write(json.dumps(rec) + "\n")

print("Wrote:", samples_path)


C:\Users\jkorm\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\Lib\importlib\__init__.py:126: UserWarning: A NumPy version >=1.23.5 and <2.3.0 is required for this version of SciPy (detected version 2.3.2)
  return _bootstrap._gcd_import(name[level:], package, level)
C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0\Lib\importlib\__init__.py:126: UserWarning: A NumPy version >=1.23.5 and <2.3.0 is required for this version of SciPy (detected version 2.3.2)
  return _bootstrap._gcd_import(name[level:], package, level)


NameError: name 'trainer' is not defined

In [ ]:
# If you hit issues on Windows, try: %pip install -U human-eval-windows && import human_eval
from human_eval.evaluation import evaluate_functional_correctness

results = evaluate_functional_correctness(str(samples_path))
print(results)
print("pass@1:", results.get("pass@1"))


Reading samples...


164it [00:00, 621.61it/s]
164it [00:00, 621.61it/s]


Running test suites...


100%|██████████| 164/164 [00:13<00:00, 11.84it/s]



Writing results to humaneval_samples.jsonl_results.jsonl...


100%|██████████| 164/164 [00:00<00:00, 15801.75it/s]

{'pass@1': np.float64(0.0)}
pass@1: 0.0


### Switching to Qwen2.5‑Coder
To try Qwen:
1. Change `MODEL_NAME` at the top to `Qwen/Qwen2.5-Coder-1.5B-Instruct`.
2. (If VRAM-limited) add `load_in_8bit=True` to the model load by exporting `BITSANDBYTES_NOWELCOME=1` and using `bnb` (already included above).
3. Increase `MAX_STEPS` and dataset size for meaningful gains.


In [1]:
import os, sys
print("Python executable:", sys.executable)
print("Kaggle kernel run type:", os.environ.get("KAGGLE_KERNEL_RUN_TYPE"))
print("Kaggle username:", os.environ.get("KAGGLE_USERNAME"))

Python executable: /usr/bin/python3
Kaggle kernel run type: Interactive
Kaggle username: None
